### PR TITLE
kitty: 0.25.0 -> 0.25.1

### PR DESCRIPTION
**Pull request by leungbk**
-------
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux (via Docker)
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by newAM — COMMENTED** [review]
I was also able to build this on `aarch64-darwin`, but an install test failed on `x86_64-linux`.

<details>
<summary>Logs...</summary>

```text
running install tests
test_os_window_size_calculation (kitty_tests.glfw.TestGLFW) ... ok
test_utf_8_strndup (kitty_tests.glfw.TestGLFW) ... ok
test_layout_operations (kitty_tests.layout.TestLayout) ... ok
test_overlay_layout_operations (kitty_tests.layout.TestLayout) ... ok
test_splits (kitty_tests.layout.TestLayout) ... ok
test_search_query_parser (kitty_tests.search_query_parser.TestSQP) ... ok
test_ip_hints (kitty_tests.hints.TestHints) ... ok
test_url_hints (kitty_tests.hints.TestHints) ... ok
test_word_trie (kitty_tests.unicode_input.TestUnicodeInput) ... ok
test_changed_center (kitty_tests.diff.TestDiff) ... ok
test_split_with_highlights (kitty_tests.diff.TestDiff) ... ok
test_backspace_wide_characters (kitty_tests.screen.TestScreen) ... ok
test_bottom_margin (kitty_tests.screen.TestScreen) ... ok
test_char_manipulation (kitty_tests.screen.TestScreen) ... ok
test_color_stack (kitty_tests.screen.TestScreen) ... ok
test_cursor_after_resize (kitty_tests.screen.TestScreen) ... ok
test_cursor_hidden (kitty_tests.screen.TestScreen) ... ok
test_cursor_movement (kitty_tests.screen.TestScreen) ... ok
test_detect_url (kitty_tests.screen.TestScreen) ... ok
test_dirty_lines (kitty_tests.screen.TestScreen) ... ok
test_draw_char (kitty_tests.screen.TestScreen) ... ok
test_draw_fast (kitty_tests.screen.TestScreen) ... ok
test_emoji_skin_tone_modifiers (kitty_tests.screen.TestScreen) ... ok
test_erase_in_screen (kitty_tests.screen.TestScreen) ... ok
test_hyperlinks (kitty_tests.screen.TestScreen) ... ok
test_key_encoding_flags_stack (kitty_tests.screen.TestScreen) ... ok
test_margins (kitty_tests.screen.TestScreen) ... ok
test_osc_52 (kitty_tests.screen.TestScreen) ... ok
test_pagerhist (kitty_tests.screen.TestScreen) ... ok
test_prompt_marking (kitty_tests.screen.TestScreen) ... ok
test_regional_indicators (kitty_tests.screen.TestScreen) ... ok
test_rep (kitty_tests.screen.TestScreen) ... ok
test_resize (kitty_tests.screen.TestScreen) ... ok
test_scrollback_fill_after_resize (kitty_tests.screen.TestScreen) ... ok
test_selection_as_text (kitty_tests.screen.TestScreen) ... ok
test_serialize (kitty_tests.screen.TestScreen) ... ok
test_sgr (kitty_tests.screen.TestScreen) ... ok
test_soft_hyphen (kitty_tests.screen.TestScreen) ... ok
test_tab_stops (kitty_tests.screen.TestScreen) ... ok
test_top_and_bottom_margin (kitty_tests.screen.TestScreen) ... ok
test_top_margin (kitty_tests.screen.TestScreen) ... ok
test_user_marking (kitty_tests.screen.TestScreen) ... ok
test_variation_selectors (kitty_tests.screen.TestScreen) ... ok
test_zwj (kitty_tests.screen.TestScreen) ... ok
test_all_kitten_names (kitty_tests.check_build.TestBuild) ... ok
test_ca_certificates (kitty_tests.check_build.TestBuild) ... skipped 'CA certificates are only tested on frozen builds'
test_exe (kitty_tests.check_build.TestBuild) ... ok
test_filesystem_locations (kitty_tests.check_build.TestBuild) ... ok
test_glfw_modules (kitty_tests.check_build.TestBuild) ... ok
test_loading_extensions (kitty_tests.check_build.TestBuild) ... ok
test_loading_shaders (kitty_tests.check_build.TestBuild) ... ok
test_pygments (kitty_tests.check_build.TestBuild) ... skipped 'Pygments is only tested on frozen builds'
test_ansi_repr (kitty_tests.datatypes.TestDataTypes) ... ok
test_color_profile (kitty_tests.datatypes.TestDataTypes) ... ok
test_historybuf (kitty_tests.datatypes.TestDataTypes) ... ok
test_line (kitty_tests.datatypes.TestDataTypes) ... ok
test_linebuf (kitty_tests.datatypes.TestDataTypes) ... ok
test_rewrap_narrower (kitty_tests.datatypes.TestDataTypes)
New buffer narrower ... ok
test_rewrap_simple (kitty_tests.datatypes.TestDataTypes)
Same width buffers ... ok
test_rewrap_wider (kitty_tests.datatypes.TestDataTypes)
New buffer wider ... ok
test_strip_csi (kitty_tests.datatypes.TestDataTypes) ... ok
test_to_color (kitty_tests.datatypes.TestDataTypes) ... ok
test_url_at (kitty_tests.datatypes.TestDataTypes) ... ok
test_utils (kitty_tests.datatypes.TestDataTypes) ... ok
test_conf_parsing (kitty_tests.options.TestConfParsing) ... ok
test_file_get (kitty_tests.file_transmission.TestFileTransmission) ... ok
test_file_put (kitty_tests.file_transmission.TestFileTransmission) ... ok
test_parse_ftc (kitty_tests.file_transmission.TestFileTransmission) ... ok
test_path_mapping_receive (kitty_tests.file_transmission.TestFileTransmission) ... ok
test_path_mapping_send (kitty_tests.file_transmission.TestFileTransmission) ... ok
test_rsync_roundtrip (kitty_tests.file_transmission.TestFileTransmission) ... ok
test_parsing_of_open_actions (kitty_tests.open_actions.TestOpenActions) ... [149 20:21:32.962874] Could not find shell to read environment
ok
test_num_users (kitty_tests.utmp.UTMPTest) ... ok
test_animation_frame_loading (kitty_tests.graphics.TestGraphics) ... ok
test_disk_cache (kitty_tests.graphics.TestGraphics) ... ok
test_gr_delete (kitty_tests.graphics.TestGraphics) ... ok
test_gr_operations_with_numbers (kitty_tests.graphics.TestGraphics) ... ok
test_gr_reset (kitty_tests.graphics.TestGraphics) ... ok
test_gr_scroll (kitty_tests.graphics.TestGraphics) ... ok
test_graphics_quota_enforcement (kitty_tests.graphics.TestGraphics) ... ok
test_image_put (kitty_tests.graphics.TestGraphics) ... ok
test_load_images (kitty_tests.graphics.TestGraphics) ... ok
test_load_png (kitty_tests.graphics.TestGraphics) ... ok
test_load_png_simple (kitty_tests.graphics.TestGraphics) ... ok
test_suppressing_gr_command_responses (kitty_tests.graphics.TestGraphics) ... ok
test_xor_data (kitty_tests.graphics.TestGraphics) ... ok
test_basic_pty_operations (kitty_tests.ssh.SSHKitten) ... ok
test_ssh_bootstrap_sh_cmd_limit (kitty_tests.ssh.SSHKitten) ... ok
test_ssh_bootstrap_with_different_launchers (kitty_tests.ssh.SSHKitten) ... ok
test_ssh_config_parsing (kitty_tests.ssh.SSHKitten) ... ok
test_ssh_connection_data (kitty_tests.ssh.SSHKitten) ... ok
test_ssh_copy (kitty_tests.ssh.SSHKitten) ... ok
test_ssh_env_vars (kitty_tests.ssh.SSHKitten) ... ok
test_ssh_leading_data (kitty_tests.ssh.SSHKitten) ... ok
test_ssh_login_shell_detection (kitty_tests.ssh.SSHKitten) ... ok
test_ssh_shell_integration (kitty_tests.ssh.SSHKitten) ... ok
test_line_edit (kitty_tests.tui.TestTUI) ... ok
test_multiprocessing_spawn (kitty_tests.tui.TestTUI) ... ok
test_mouse_selection (kitty_tests.mouse.TestMouse) ... ok
test_case_insensitive (kitty_tests.choose.TestMatcher) ... ok
test_delimiter (kitty_tests.choose.TestMatcher)
Test using a custom line delimiter ... ok
test_filtering (kitty_tests.choose.TestMatcher)
Non matching entries must be removed ... ok
test_marking (kitty_tests.choose.TestMatcher)
Marking of matched characters ... ok
test_positions (kitty_tests.choose.TestMatcher)
Output of positions ... ok
test_scoring (kitty_tests.choose.TestMatcher)
Scoring algorithm ... ok
test_threading (kitty_tests.choose.TestMatcher)
Test matching on a large data set with different number of threads ... ok
test_charsets (kitty_tests.parser.TestParser) ... ok
test_csi_code_rep (kitty_tests.parser.TestParser) ... ok
test_csi_codes (kitty_tests.parser.TestParser) ... ok
test_dcs_codes (kitty_tests.parser.TestParser) ... ok
test_deccara (kitty_tests.parser.TestParser) ... ok
test_desktop_notify (kitty_tests.parser.TestParser) ... ok
test_esc_codes (kitty_tests.parser.TestParser) ... ok
test_graphics_command (kitty_tests.parser.TestParser) ... ok
test_osc_codes (kitty_tests.parser.TestParser) ... ok
test_oth_codes (kitty_tests.parser.TestParser) ... ok
test_pending (kitty_tests.parser.TestParser) ... ok
test_sc81t (kitty_tests.parser.TestParser) ... ok
test_simple_parsing (kitty_tests.parser.TestParser) ... ok
test_box_drawing (kitty_tests.fonts.Rendering) ... [149 20:21:36.274432] Failed to find and load fontconfig
```
</details>

File: pkgs/applications/terminal-emulators/kitty/default.nix [hunk 1]
```diff
@@ -78,12 +78,6 @@ buildPythonApplication rec {
   outputs = [ "out" "terminfo" "shell_integration" ];
 
   patches = [
-    # Required to get `test_ssh_env_vars` to pass.
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/applications/terminal-emulators/kitty/default.nix:82]: Please remove the unused fetchpatch from inputs.

**Review by malob — CHANGES_REQUESTED** [review]
Looks like some commits have been added upstream to address some of the issues. I suggest the following changes:

```patch
diff --git a/pkgs/applications/terminal-emulators/kitty/default.nix b/pkgs/applications/terminal-emulators/kitty/default.nix
index 8497153acbb..cf5acdd5dd8 100644
--- a/pkgs/applications/terminal-emulators/kitty/default.nix
+++ b/pkgs/applications/terminal-emulators/kitty/default.nix
@@ -79,16 +79,23 @@ buildPythonApplication rec {
 
   patches = [
     # Fix to ensure that files in tar files used by SSH kitten have write permissions.
-    ./tarball-restore-write-permissions.patch
+    (fetchpatch {
+      name = "fix-tarball-file-permissions.patch";
+      url = "https://github.com/kovidgoyal/kitty/commit/8540ca399053e8d42df27283bb5dd4af562ed29b.patch";
+      sha256 = "sha256-y5w+ritkR+ZEfNSRDQW9r3BU2qt98UNK7vdEX/X+mKU=";
+    })
 
     # Needed on darwin
 
     # Gets `test_ssh_shell_integration` to pass for `zsh` when `compinit` complains about
     # permissions.
     ./zsh-compinit.patch
-    # Skip `test_ssh_login_shell_detection` in some cases, build users have their shell set to
-    # `/sbin/nologin` which causes issues.
-    ./disable-test_ssh_login_shell_detection.patch
+    # Skip login shell detection when login shell is set to nologin
+    (fetchpatch {
+      name = "skip-login-shell-detection-for-nologin.patch";
+      url = "https://github.com/kovidgoyal/kitty/commit/27906ea853ce7862bcb83e324ef80f6337b5d846.patch";
+      sha256 = "sha256-Zg6uWkiWvb45i4xcp9k6jy0R2IQMT4PXr7BenzZ/md8=";
+    })
     # Skip `test_ssh_bootstrap_with_different_launchers` when launcher is `zsh` since it causes:
     # OSError: master_fd is in error condition
     ./disable-test_ssh_bootstrap_with_different_launchers.patch
```

**Comment by leungbk** [comment]
By patching `fontconfig.c`, I am able to get this to build with Nix in a Docker container. I am not sure if this is the best way of making this work.

I know nothing about Docker, and very little about Kitty, so I don't know if I can easily test whether https://github.com/kovidgoyal/kitty/blob/cc4f1c0a249f56a837680ccc2c1a43d69ba9925e/kitty/desktop.c#L40 and https://github.com/kovidgoyal/kitty/blob/cc4f1c0a249f56a837680ccc2c1a43d69ba9925e/kitty/desktop.c#L116 need to also be patched.

**Review by newAM — APPROVED** [review]
Builds and runs on `x86_64-linux` now.

I'll let someone else comment on the `libfontconfig` patch.

**Comment by sandangel** [comment]
It's working for me,

```nix
  home.packages = with pkgs; [
    (let src = fetchgit {
      url = "https://github.com/leungbk/nixpkgs";
      sparseCheckout = ''
        pkgs/applications/terminal-emulators/kitty
      '';
      rev = "f3c611217c8973317b671c95274bcb5e86de0527";
      sha256 = "sha256-6a3FntHp8VNTe4ZJEZoOM0chHd40jUA7gjEafZimD1s=";
    };
    in callPackage "${src}/pkgs/applications/terminal-emulators/kitty" {
      inherit (darwin.apple_sdk.frameworks) Cocoa CoreGraphics Foundation IOKit Kernel OpenGL;
    })
  ];
```

I'm using the aarch64-linux VM on mac m1 

Not sure why did I have to add `darwin.apple_sdk` though.

File: pkgs/applications/terminal-emulators/kitty/libfontconfig.patch [hunk 2]
```diff
@@ -0,0 +1,22 @@
+diff --git a/kitty/fontconfig.c b/kitty/fontconfig.c
+index 9710f210..803ce242 100644
+--- a/kitty/fontconfig.c
++++ b/kitty/fontconfig.c
+@@ -80,13 +80,7 @@ static struct {
+ static void
+ load_fontconfig_lib(void) {
+         const char* libnames[] = {
+-#if defined(_KITTY_FONTCONFIG_LIBRARY)
```
[the remark this answers is not in the document]
Comment (leungbk) [reply pkgs/applications/terminal-emulators/kitty/libfontconfig.patch]: Yup, read the code a bit more carefully and determined it was an upstream bug. The relevant patches are now attached.

**Comment by leungbk** [comment]
> It's working for me,
> 
> ```nix
>   home.packages = with pkgs; [
>     (let src = fetchgit {
>       url = "https://github.com/leungbk/nixpkgs";
>       sparseCheckout = ''
>         pkgs/applications/terminal-emulators/kitty
>       '';
>       rev = "f3c611217c8973317b671c95274bcb5e86de0527";
>       sha256 = "sha256-6a3FntHp8VNTe4ZJEZoOM0chHd40jUA7gjEafZimD1s=";
>     };
>     in callPackage "${src}/pkgs/applications/terminal-emulators/kitty" {
>       inherit (darwin.apple_sdk.frameworks) Cocoa CoreGraphics Foundation IOKit Kernel OpenGL;
>     })
>   ];
> ```
> 
> I'm using the aarch64-linux VM on mac m1
> 
> Not sure why did I have to add `darwin.apple_sdk` though.

@sandangel see https://nixos.org/guides/nix-pills/callpackage-design-pattern.html. You're calling the `kitty/default.nix` with arguments set by default to their corresponding values in `all-packages.nix`. Your darwin overrides are needed because although `kitty/default.nix` takes `Cocoa`, `CoreGraphics`, etc. arguments, those arguments are not defined at the top level `all-packages.nix`; however, there is a `darwin` value in that package set that has the right shape (`apple_sdk` attr, inside of which lies a `frameworks` attr, etc...)

The reason you don't ordinarily need to think about this in your config files is that you're typically installing `pkgs.kitty` as it is defined in `all-packages.nix`, where the overrides are already specified, whereas in your example you're  defining your own derivation.

**Comment by sandangel** [comment]
oh nice, thanks for helping me understand that. 😃

**Pull request merged**